MonoZelda/Link/Projectiles/ProjectileManager.cs

```diff
@@ -1,213 +1,93 @@
+                break;
         }
     }
 
-    private void setupProjectile(ProjectileType equippedProjectile)
+    public void UseSword()
     {
-        if (!hasRequiredResources(equippedProjectile))
-        {
-            return;
-        }
-        
-        if (equippedProjectile == ProjectileType.WoodenSword)
+        IProjectile sword;
+        if (PlayerState.Health == PlayerState.MaxHealth && ActiveProjectiles.ContainsKey(ProjectileType.WoodenSwordBeam) == false)
         {
-            isSwordAttack = true;
+            sword = projectileFactory.GetProjectileObject(ProjectileType.WoodenSwordBeam,collisionController);
+            ActiveProjectiles.Add(ProjectileType.WoodenSwordBeam, sword);
         }
-        else if (equippedProjectile == ProjectileType.WoodenSwordBeam)
-        {
-            isSwordBeamAttack = true;
-        }
-
-        deductResources(equippedProjectile);
-        projectileFired = true;
-        projectileDict.Enabled = true;
-        if (equippedProjectile != ProjectileType.Bomb)
+        else
         {
-            activateHitbox = true;
+            sword = projectileFactory.GetProjectileObject(ProjectileType.WoodenSword,collisionController);
+            ActiveProjectiles.Add(ProjectileType.WoodenSword, sword);
         }
+        sword.Setup();
     }
 
-    public void equipProjectile(Keys pressedKey)
+    public void FireProjectile(ProjectileType projectileType)
     {
-        if (keyProjectileMap.TryGetValue(pressedKey, out ProjectileType newProjectile))
+        if (SingleInstanceProjectiles.Contains(projectileType))
         {
-            EquippedProjectile = newProjectile;
-        }
-    }
-
-    public void destroyProjectile()
-    {
-        itemFired.FinishProjectile();
-    }
-
-    public void useSword(PlayerSpriteManager player)
-    {
-        itemFired = projectileFactory.GetProjectileObject(ProjectileType.WoodenSword, player);
-        setupProjectile(ProjectileType.WoodenSword);
-    }
-
-    public void useSwordBeam(PlayerSpriteManager player)
-    {
-        itemFired = projectileFactory.GetProjectileObject(ProjectileType.WoodenSwordBeam, player);
-        setupProjectile(ProjectileType.WoodenSwordBeam);
-    }
-
-    public void fireEquippedProjectile(PlayerSpriteManager player)
-    {
-        if (!hasRequiredResources(EquippedProjectile) || EquippedProjectile == ProjectileType.None)
-        {
-            return;
-        }
-        //prevent candle use in more than one room
-        if (EquippedProjectile == ProjectileType.CandleBlue && PlayerState.IsCandleUsed)
-        {
-            return; 
+            if (ActiveProjectiles.ContainsKey(projectileType) == false)
+            {
+                IProjectile projectile = projectileFactory.GetProjectileObject(projectileType, collisionController);
+                ActiveProjectiles.Add(projectileType, projectile);
+                DeductPlayerResources(projectileType);
+                projectile.Setup();
+            }
```

Comment:
Big improvement in abstracting projectile types in order to make `ProjectileManager`'s job easier